src/components/Microsoft.DotNet.UpgradeAssistant.Extensions/ExtensionManager.cs

```diff
@@ -0,0 +1,79 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.DotNet.UpgradeAssistant.Extensions
+{
+    internal sealed class ExtensionManager : IDisposable, IEnumerable<ExtensionInstance>
+    {
+        private readonly Lazy<IEnumerable<ExtensionInstance>> _extensions;
+
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Design", "CA1031:Do not catch general exception types", Justification = "Loading an extension should not propogate any exceptions.")]
+        public ExtensionManager(
+            IEnumerable<IExtensionLoader> loaders,
+            IOptions<ExtensionOptions> options,
+            ILogger<ExtensionManager> logger)
+        {
+            _extensions = new Lazy<IEnumerable<ExtensionInstance>>(() =>
+            {
+                var list = new List<ExtensionInstance>();
+
+                foreach (var path in options.Value.ExtensionPaths)
+                {
+                    if (LoadExtension(path) is ExtensionInstance extension)
+                    {
+                        list.Add(extension);
+                    }
+                    else
+                    {
+                        logger.LogWarning("Could not load extension from {Path}", path);
+                    }
+                }
+
+                logger.LogInformation("Loaded {Count} extensions", list.Count);
+
+                return list;
+            });
+
+            ExtensionInstance? LoadExtension(string path)
+            {
+                foreach (var loader in loaders)
+                {
+                    try
+                    {
+                        if (loader.LoadExtension(path) is ExtensionInstance instance)
```

Comment:
Not sure if this is worth changing, but rather than trying all the loaders, `IExtensionLoader` could have an API for determining whether a path matches their expected format (*.zip, manifest.json, etc.) and you could try loading the extension with only the loaders that indicate they can handle the path.

Reply:
Created an issue to track that change